### PR TITLE
feat: Cluster terminal manifest edit, debug mode

### PR DESCRIPTION
**Pull request by milindk8**
# Description

In the cluster terminal, now we are providing an option to the user to switch between debug node mode and utility mode. And also allows users to edit the manifest of the pod.

Fixes [# AB2348](https://dev.azure.com/DevtronLabs/Devtron/_boards/board/t/OSS%20adoption/Stories/?workitem=2348)

Screenshot 2023-05-04 at 9 46 17 AM

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
All testing is done manually

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,6 +19,7 @@
         "flexsearch": "^0.6.32",
         "http-proxy-middleware": "^2.0.4",
         "marked": "^4.0.18",
+        "mock-socket": "^9.2.1",
```
Comment (vivek-devtron) [inline package.json:22]: this should be in dev dependency. It is used for testing. right?

File: src/components/cluster/cluster.util.ts [hunk 2]
```diff
@@ -48,3 +49,11 @@ export function getClusterTerminalParamsData(
         selectedShell: _selectedShell,
     }
 }
+
+export const createTaintsList = (list: any[], nodeLabel: string): Map<string, NodeTaintType[]> => {
+    const taints = new Map<string, NodeTaintType[]>()
+    list?.forEach((node) => {
```
Comment (sohel-rp) [inline src/components/cluster/cluster.util.ts:55]: try to use reduce

File: src/components/ClusterNodes/constants.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import { POD_LINKS } from '../v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/node.type'
```
Comment (sohel-rp) [inline src/components/ClusterNodes/constants.ts:1]: why is this in type?

File: src/components/ClusterNodes/ClusterManifest.tsx [hunk 4]
```diff
@@ -3,53 +3,162 @@ import CodeEditor from '../CodeEditor/CodeEditor'
...
                 })
         } else {
             setResourceMissing(true)
             setLoading(false)
+            setManifestAvailable(false)
         }
     }, [terminalAccessId])
 
+    useEffect(() => {
+        const regex = /^.*apiVersion:/
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/ClusterManifest.tsx:54]: Please move it to constants file

File: src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/terminal.type.ts [hunk 5]
```diff
@@ -2,9 +2,10 @@ import React from 'react'
 import { OptionsOrGroups, GroupBase, ActionMeta, StylesConfig } from 'react-select'
 import { SelectComponents } from 'react-select/dist/declarations/src/components'
 import { SocketConnectionType } from '../node.type'
+import { EDIT_MODE_TYPE, TERMINAL_WRAPPER_COMPONENT_TYPE } from './constants'
```
Comment (vivek-devtron) [inline src/components/v2/appDetails/k8Resource/nodeDetail/NodeDetailTabs/terminal/terminal.type.ts:5]: Please create types in PascalCase

File: src/components/ClusterNodes/NodeDetails.tsx [hunk 6]
```diff
@@ -171,39 +169,42 @@ export default function NodeDetails({ imageList, isSuperAdmin, namespaceList }:
...
-            search: _searchParam,
-        })
     }
 
     const renderNodeDetailsTabs = (): JSX.Element => {
+        const cursorValue = isNodeAuto ? 'cursor-not-allowed' : 'cursor'
         return (
             <ul role="tablist" className="tab-list">
-                <li className="tab-list__tab pointer" data-tab-index="0" onClick={changeNodeTab}>
+                <li className={`tab-list__tab pointer ${cursorValue}`} data-tab-index="0" onClick={changeNodeTab}>
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/NodeDetails.tsx:194]: we should remove pointer class

File: src/components/ClusterNodes/NodeDetails.tsx [hunk 7]
```diff
@@ -171,39 +169,42 @@ export default function NodeDetails({ imageList, isSuperAdmin, namespaceList }:
...
                     <div className={`mb-6 fs-13 tab-hover${selectedTabIndex == 0 ? ' fw-6 active' : ' fw-4'}`}>
                         {NODE_DETAILS_TABS.summary}
                     </div>
                     {selectedTabIndex == 0 && <div className="node-details__active-tab" />}
                 </li>
-                <li className="tab-list__tab pointer" data-tab-index="1" onClick={changeNodeTab}>
+                <li className={`tab-list__tab pointer ${cursorValue}`} data-tab-index="1" onClick={changeNodeTab}>
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/NodeDetails.tsx:200]: same here

File: src/components/ClusterNodes/NodeDetails.tsx [hunk 8]
```diff
@@ -171,39 +169,42 @@ export default function NodeDetails({ imageList, isSuperAdmin, namespaceList }:
...
                     <div className={`mb-6 flexbox fs-13 tab-hover${selectedTabIndex == 1 ? ' fw-6 active' : ' fw-4'}`}>
                         <Edit className="icon-dim-16 mt-2 mr-5 edit-yaml-icon" />
                         {NODE_DETAILS_TABS.yaml}
                     </div>
                     {selectedTabIndex == 1 && <div className="node-details__active-tab" />}
                 </li>
-                <li className="tab-list__tab pointer" data-tab-index="2" onClick={changeNodeTab}>
+                <li className={`tab-list__tab pointer ${cursorValue}`} data-tab-index="2" onClick={changeNodeTab}>
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/NodeDetails.tsx:207]: same here

**Review by shivani170 — APPROVED** [review]

**Pull request merged**